fix(ivy): don't detect changes on detached child embedded views

crisbeto opened the pull request:
Fixes Ivy detecting changes inside child embedded views, even though they're detached.

Note that there's one subtlety here: I made the changes inside `refreshDynamicEmbeddedViews` rather than `refreshView`, because we support detecting changes on a detached view (evidenced by a couple of unit tests), but only if it's triggered directly from the view's `ChangeDetectorRef`, however we shouldn't be detecting changes in the detached child view when something happens in the parent.

**Note:** I had to bump the size limit, because it looks like master was right on the edge before these changes. I haven't introduced any new imports and the new logic is only a couple of lines so I doubt that these changes would've added too much to the size.

Fixes #34816.

pkozlowski-opensource commented on packages/core/test/acceptance/change_detection_spec.ts:
```diff
@@ -70,6 +68,49 @@ describe('change detection', () => {
     });
 
+    it('should not detect changes in child embedded views while they are detached', () => {
+      const counters = {outsideTemplate: 0, insideTemplate: 0};
```
nit: personally I would name those to be named `componentView` and `embeddedView` as usage of the "template" in those names made me think that there are 2 `<ng-template>` elements here. It might be just me, though...

pkozlowski-opensource commented on packages/core/test/acceptance/change_detection_spec.ts:
```diff
@@ -70,6 +68,49 @@ describe('change detection', () => {
+
+      TestBed.configureTestingModule({declarations: [App, ViewManipulation]});
+      const fixture = TestBed.createComponent(App);
+      const vm: ViewManipulation = fixture.debugElement.childNodes[2].references['vm'];
```
super-nit / personal preference: in cases like this I usually do `debugElement.query(By.directive(ViewManipulation)).injector.get(ViewManipulation)` as I find this approach less "brittle". (BTW, for quite a bit of time I wanted to have a better API to do just this in `TestBed` but this is a separate story....).

crisbeto replied:
EDIT: I tried it out and `query` didn't manage to find the directive for some reason. Maybe because I'm trying to access it before the first change detection run?

Sure, I don't feel strongly about it. I did it this way, because the other unit tests were doing the same.

pkozlowski-opensource commented on packages/core/test/acceptance/change_detection_spec.ts:
```diff
@@ -70,6 +68,49 @@ describe('change detection', () => {
+      fixture.detectChanges();
+      expect(counters).toEqual({outsideTemplate: 2, insideTemplate: 2});
+
+      viewRef.detach();
```
yep, yep, yep.... corner case but definitively possible given our APIs....

pkozlowski-opensource commented on packages/core/test/acceptance/change_detection_spec.ts:
```diff
@@ -70,6 +68,49 @@ describe('change detection', () => {
+
+      expect(counters).toEqual({outsideTemplate: 4, insideTemplate: 3});
+    });
+
```
question / idea: do you think it would make sense to add another test where we insert a component view (and not only an embedded view) to make sure that we cover all the cases?

crisbeto replied:
Not sure what you mean by a component view. Do you mean through `ViewContainerRef.createComponent`?

pkozlowski-opensource replied:
yep, `ViewContainerRef.createComponent`

crisbeto replied:
Done.

pkozlowski-opensource commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1525,7 +1525,10 @@ function refreshDynamicEmbeddedViews(lView: LView) {
         const embeddedTView = embeddedLView[TVIEW];
         ngDevMode && assertDefined(embeddedTView, 'TView must be allocated');
-        refreshView(embeddedLView, embeddedTView, embeddedTView.template, embeddedLView[CONTEXT] !);
+        if (viewAttachedToChangeDetector(embeddedLView)) {
```
thinking out loud: I wonder if it wouldn't be better to move this check to the `refreshView` so we don't need to read `lView[FLAGS]` here _and_ inside `refreshView` for the most common case. WDYT?

crisbeto replied:
I commented on this in the PR description: I initially had it inside of `refreshView`, but that means that we can't detect changes on detached views which we seems to support (there are two unit tests for it).

pkozlowski-opensource replied:
Ohhhh, I see, this makes sense, you are right! Thnx for clarifying!

pkozlowski-opensource reviewed (approved):
LGTM overall, just left few comments / questions that I would like have your thoughts on.

crisbeto reviewed (commented):
I've addressed/replied to all of the feedback @pkozlowski-opensource.

crisbeto commented:
I've added the extra unit test.

crisbeto committed:
fix(ivy): don't detect changes on detached child embedded views

Fixes Ivy detecting changes inside child embedded views, even though they're detached.

Note that there's on subtlety here: I made the changes inside `refreshDynamicEmbeddedViews` rather than `refreshView`, because we support detecting changes on a detached view (evidenced by a couple of unit tests), but only if it's triggered directly from the view's `ChangeDetectorRef`, however we shouldn't be detecting changes in the detached child view when something happens in the parent.

Fixes #34816.

pkozlowski-opensource reviewed (approved):
LGTM. The size increase is unfortunate but I believe that it is mostly because of the creeping size increase from the previous PRs.

AndrewKushnir commented:
[Presubmit](http://test/OCL:291382586:BASE:291382654:1579885557647:7fc18c5f)

IgorMinar reviewed (approved):
lgtm  - global approval (the size diff is not caused by this change)
